Keep splitPostGRA stores in right block for acmp fast path

hzongaro opened the pull request:
When `splitPostGRA` splits a block, it might insert register stores and stores to temporaries at the end of the original block at the split point.  When `splitPostGRA` is called for `fastpathAcmpHelper`, the original block contains the `acmp` helper call that will still need to be moved into a separate "call" block, while any register stores or temporary stores that do not involve the call to `acmp` helper will need to remain in the original block.  Move any treetops, register stores or stores
for the call to the `acmp` helper to the end of the original block so that the trees associated with the call can be split into the new "call" block without pulling any unrelated stores and register stores into the new block.

Also, register dependencies that `splitPostGRA` adds to the end of the "call" block should always be copied to the `ifacmpeq` node that's added to the end of the original block as that represents an edge to the same "target" block.

This fixes a failure with value types enabled running the vmTest_2 test, UnsafeArrayGetTest.

r30shah commented on runtime/compiler/codegen/J9CodeGenerator.cpp:
```diff
@@ -251,6 +251,48 @@ J9::CodeGenerator::fastpathAcmpHelper(TR::Node *node, TR::TreeTop *tt, const boo
+         TR::TreeTop* nextTT = afterTT->getNextTreeTop();
+         TR::ILOpCodes op = afterTT->getNode()->getOpCodeValue();
+
+         if ((op == TR::iRegStore || op == TR::istore) && afterTT->getNode()->getFirstChild() == node)
```
Why we move the acmp call tree top just after node whose child is node? I think this routine can be simpler, what you need to do is move the `acmp` call helper at the end of the `prevBlock` . You might just need to get the last real tree-top and add it after it as you have just split the blocks I do not think we would have more than one exit edge from the `prevBlock`.

hzongaro replied:
@r30shah Rahil, I'm not sure I'm understanding your comment, so let me show an example of what's happening.  Here are the trees after the call to `splitPostGRA`:

```
n142n     treetop                                                                             
n141n       icall  jitAcmpHelper[#99  helper Method] [flags 0x400 0x0 ]                       
n139n         ==>aload
n140n         ==>aload
n340n     bRegStore ecx                                                                       
n14n        ==>bloadi
n343n     bRegStore edx                                                                       
n31n        ==>bloadi
n346n     bRegStore esi                                                                       
n48n        ==>bloadi
n349n     bRegStore eax                                                                       
n65n        ==>bloadi
n352n     bRegStore r14d                                                                      
n82n        ==>bloadi
n355n     istore  <temp slot 24>[#371  Auto] [flags 0x3 0x0 ]                                 
n354n       b2i                                                                               
n99n          ==>bloadi
n359n     istore  <temp slot 25>[#372  Auto] [flags 0x3 0x0 ]                                 
n358n       b2i                                                                               
n116n         ==>bloadi
n363n     istore  <temp slot 26>[#373  Auto] [flags 0x3 0x0 ]                                 
n362n       b2i                                                                               
n133n         ==>bloadi
n366n     istore  <temp slot 27>[#374  Auto] [flags 0x3 0x0 ]                                 
n141n       ==>icall
n338n     BBEnd </block_2> ===== ()
```

I'm trying to move the `TR::TreeTop`s containing nodes n142n and n366n.  I hadn't wanted to make any assumptions about how many stores there might be for the value of the `icall` or whether they would appear after all of the other stores.

r30shah replied:
Thanks a lot for the example. So when you split the block, is `n141n` referenced after split point?

r30shah replied:
Okk, Sorry, now it is clear to me.  So far it seems good to me, let me give it another look.

a7ehuo commented on runtime/compiler/codegen/J9CodeGenerator.cpp:
```diff
@@ -251,6 +251,48 @@ J9::CodeGenerator::fastpathAcmpHelper(TR::Node *node, TR::TreeTop *tt, const boo
    TR::Block* prevBlock = tt->getEnclosingBlock();
    TR::Block* targetBlock = prevBlock->splitPostGRA(tt->getNextTreeTop(), cfg, true, NULL);
+
+   // Move call tree to end of block, along with any stores of the result of the call, if they're
```
I wonder if the diagram on the above should be updated with `GlRegDeps()` to  show how the register dependency looks like between the split blocks and some sort of annotation that mentions the call tree needs to be at the end of the block.
https://github.com/eclipse/openj9/blob/bd3aac0b531b17e22787e8f8cf484522859c8f3f/runtime/compiler/codegen/J9CodeGenerator.cpp#L197-L219

hzongaro replied:
Thanks for suggesting that this should show the register dependencies; I'll do that.

As for showing that the call tree has to appear at the end of the block, the difficulty is that this diagram shows where the `acmphelper` call ends up after the call to `split`.  It's only at the end of the previous block just before the call `split`, so I would have to add another diagram showing the intermediate state.  I'm not sure it's worth it to add another diagram.  Thoughts?

r30shah replied:
IMO a detailed explanation here before moving the treetops should suffice the intermediate state. Adding CFG diagram for intermediate state in the beginning would complicate.

a7ehuo replied:
A detailed explanation would do.

a7ehuo commented on runtime/compiler/codegen/J9CodeGenerator.cpp:
```diff
@@ -251,6 +251,48 @@ J9::CodeGenerator::fastpathAcmpHelper(TR::Node *node, TR::TreeTop *tt, const boo
+            {
+            if (trace)
+               {
+               traceMsg(comp, "Moving treetop containing node n%dn [%p] associated with acmp helper call to end of prevBlock in preparation of final block split\n", afterTT->getNode()->getGlobalIndex(), afterTT->getNode());
```
Should this trace be updated to differentiate from the trace message at line L265 to show this one is more specific to store nodes?

a7ehuo reviewed (commented):
Would `lowerArrayStoreCHK` also have similar split issue?

hzongaro commented:
Annabelle @a7ehuo, you asked,

> Would `lowerArrayStoreCHK` also have similar split issue as `fastpathAcmpHelper`?

I don't think it can.  After calling `splitPostGRA`, [`lowerArrayStoreCHK` appends an `ifacmpeq` and `NULLCHK`](https://github.com/eclipse/openj9/blob/8c76311cdcccaa6def680d362df418eca6ae9325/runtime/compiler/codegen/J9CodeGenerator.cpp#L1628-L1663), and then calls `split` on the original block at the point of the `NULLCHK`.  So there shouldn't be any stores that were produced by the call to `splitPostGRA` after the `NULLCHK`.

r30shah commented on runtime/compiler/codegen/J9CodeGenerator.cpp:
```diff
@@ -251,6 +251,48 @@ J9::CodeGenerator::fastpathAcmpHelper(TR::Node *node, TR::TreeTop *tt, const boo
+   // not already at the end.  The call has to be split into its own block, but splitPostGRA above
+   // might have left register stores and stores into temporaries that must stay with the previous block
+   TR::TreeTop* prevBlockExit = prevBlock->getExit();
+   TR::TreeTop* afterTT = tt->getNextTreeTop();
```
This is used as `iterator` mainly to walk through treetops. How about using `iterTT`?

r30shah commented on runtime/compiler/codegen/J9CodeGenerator.cpp:
```diff
@@ -251,6 +251,48 @@ J9::CodeGenerator::fastpathAcmpHelper(TR::Node *node, TR::TreeTop *tt, const boo
+
+      // Remove TreeTop for call node
+      tt->unlink(false);
+      TR::TreeTop* callTreesTail = tt;
```
This  is the last real tree-top in the later created `callBlock` , may be we can rename this to reflect that.

r30shah reviewed (approved):
Thanks a lot for clearing some of the doubts I had offline. Overall Changes looks good to me, just minor suggestions I have posted in the review.

hzongaro commented:
Annabelle @a7ehuo, Rahil @r30shah, I've addressed your review comments in commit 6097e70a855701546db5537c4a4e6d3dcb4ed941.  If you're OK with those changes, I will squash the commits; otherwise, I'll address any additional comments you might have.

r30shah reviewed (approved):
LGTM

a7ehuo commented:
LGTM

hzongaro commented:
Just squashed down to one commit - ed8362321085059d28e7d7e0d1ee86d14f1e6362

hzongaro commented:
Daryl @0xdaryl, may I ask you to review this change?

0xdaryl commented:
Jenkins test sanity all jdk11

0xdaryl commented:
The testing results were successful even though the PR checks did not get updated (there was an infra issue).  Here are the results -> https://ci.eclipse.org/openj9/view/Pull%20Requests/job/PullRequest-OpenJ9/5017/ , although this link will likely be extinct shortly.
